fw/ringbuffer.c

```diff
@@ -0,0 +1,248 @@
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program; if not, write to the Free Software Foundation, Inc., 59
+ * Temple Place - Suite 330, Boston, MA 02111-1307, USA.
+ */
+
+#include "ringbuffer.h"
+#include "lib/str.h"
+#include <linux/types.h>
+#include <asm/io.h>
+#include <linux/log2.h>
+#include <linux/cpumask.h>
+
+typedef struct {
+	TfwRingbuffer *rb;
+	struct page *pg;
+	atomic_t unmapped;
+} TfwRingbufferArea;
+
+#define AREA_SIZE(size) (size + TFW_RINGBUFFER_DATA_OFFSET)
+#define RBA_AREA_SIZE(rba) AREA_SIZE((rba)->rb->size)
+
+static DEFINE_PER_CPU(TfwRingbufferArea, rb_area);
+static bool proc_file_is_open;
+
+int
+tfw_ringbuffer_write(TfwStr **strs, unsigned int count)
+{
+	TfwRingbufferArea *rba = this_cpu_ptr(&rb_area);
+	TfwRingbuffer *rb = rba->rb;
+	unsigned int i;
+	u64 head, tail;
+	u32 full_size = 0;
+
+	if (atomic_read(&rba->unmapped))
+		return -EAGAIN;
+
+	preempt_disable();
+
+	head = rb->head;
+	tail = smp_load_acquire(&rb->tail);
+
+	for (i = 0; i < count; ++i)
+		full_size += strs[i]->len;
+
+	if (head - tail + full_size >= rb->size) {
+		preempt_enable();
+		return -ENOMEM;
+	}
+
+	for (i = 0; i < count; ++i) {
+		TfwStr *c, *end;
+
+		TFW_STR_FOR_EACH_CHUNK(c, strs[i], end) {
+			char *data = c->data;
+			u32 len = c->len;
+
+			while (len) {
+				u32 mhead, cur_len;
+
+				mhead = head & rb->mask;
+				cur_len = min(len, rb->size - mhead);
+
+				memcpy_fast(rb->data + mhead, data, cur_len);
+				data += cur_len;
+				head += cur_len;
+				len -= cur_len;
+			}
+		}
+	}
+
+	smp_store_release(&rb->head, head);
+	smp_mb();
+
+	preempt_enable();
+
+	return 0;
+}
+
+int
+tfw_ringbuffer_file_open(struct inode *ino, struct file *filp)
+{
+	if (proc_file_is_open)
+		return -EBUSY;
+	proc_file_is_open = true;
+	return 0;
+}
+
+int
+tfw_ringbuffer_file_close(struct inode *ino, struct file *filp)
+{
+	int cpu;
+	proc_file_is_open = false;
+	for_each_online_cpu(cpu) {
+		TfwRingbufferArea *rba = per_cpu_ptr(&rb_area, cpu);
+		atomic_set(&rba->unmapped, 1);
+	}
+	return 0;
+}
+
+int
+tfw_ringbuffer_file_mmap(struct file *filp, struct vm_area_struct *vma)
+{
+	TfwRingbufferArea *rba, *this_rba = this_cpu_ptr(&rb_area);
+	unsigned long pfn, size, area_size, area_pages;
+	int cpu;
+
+	area_size = RBA_AREA_SIZE(this_rba);
+	size = vma->vm_end - vma->vm_start;
+	if (size > area_size)
+		return -EINVAL;
+
+	area_pages = area_size / PAGE_SIZE;
+
+	cpu = vma->vm_pgoff / area_pages;
+	if (cpu >= num_online_cpus())
+		return -EINVAL;
+
+	rba = per_cpu_ptr(&rb_area, cpu);
+	pfn = page_to_pfn(rba->pg);
+
+	if (remap_pfn_range(vma, vma->vm_start, pfn, size, vma->vm_page_prot))
+		return -EAGAIN;
+
+	if (size == area_size)
+		atomic_set(&rba->unmapped, 0);
+
+	return 0;
+}
```

Comment:
So we're going to have per-cpu file, right? Could you please comment the function about this consideration? Basically having a minimal documentation in the source, like I referened in the linux kernel, would be useful

Reply:
No, just one file with multiple mapped per-cpu areas.

There is no appropriate munmap field in `struct file_operations`. I could set vma->vm_ops and catch unmap event, but I think it would be better to move this field to the Ringbuffer structure, so a client could change it when it prepared to read.